### PR TITLE
Multipath descriptors

### LINKED ISSUE #469: Multipath descriptors support
Multipath descriptors were proposed as a way to optimize the descriptor backup for the common case of having two derivation chains (one for receiving addresses, one for change).

A BIP was proposed a couple months ago: https://github.com/bitcoin/bips/pull/1354. There is an implementation in Bitcoin Core at https://github.com/bitcoin/bitcoin/pull/22838.

It was previously discussed at length in https://github.com/bitcoin/bitcoin/issues/17190.

I think it would be helpful for us to support them. What do you think?

### PR DESCRIPTION
**Pull request by darosior**
This implements [BIP389 multipath descriptors](https://github.com/bitcoin/bips/pull/1354). Fixes #469.

### CODE REVIEW

**Comment by sanket1729** [comment]
I briefly looked at the code and I like the overall approach for the given BIP. Overall, a strong concept and approach ACK. Just two things:

1) I think Conversion to `DefininateDescriptorKey` should fail when there is MultiPath descriptor
2) We can add `MiniscriptKey::num_derivation_paths`/`MiniscriptKey::num_paths` method to `MiniscriptKey` trait. When converting `Terminal` to `Miniscript`, we check whether all keys have the same `num_derivation_paths`/`num_pahts`, otherwise an error. For all existing types that implement `MiniscriptKey`, this can return 1. I am undecided/unsure whether we want a default impl returning `1` or if we should force all implementers to write the function.
Lastly, for taproot descriptors, it would require checking the same thing in all miniscripts and internal keys in all the initialization constructors.

**Comment by darosior** [comment]
> On another note, I don't like the BIP changes in KEY expressions, but I have not formed an opinion about it. Previously, we could reason about legality of KEY expressions only based on top level contexts, but with this a KEY expression can invalidate another KEY expression in some other part of descriptor

I very much agree. This is [the feedback i gave to Achow](https://github.com/bitcoin/bips/pull/1354#discussion_r985316649) when implementing this.

I was kinda waiting for an answer from him about this, but looks like it won't come. I'll address your feedback.

**Comment by darosior** [comment]
Implemented your suggestions and removed the WIP label from this PR. Having a `num_der_paths` method on `MiniscriptKey` allowed to check for inconsistent multipath keys directly in `Descriptor<Pk>::from_str`.

> I am undecided/unsure whether we want a default impl returning 1 or if we should force all implementers to write the function

Rather returning `0` by default? In any case i don't think there is a sane default for this. It really depends on the implementer. So i went for no default implementation: should we want to, i *think* we can always introduce one later on without breaking backward compatibility.

File: src/descriptor/key.rs [hunk 1]
```diff
@@ -314,6 +314,44 @@ impl DescriptorSecretKey {
...
         Ok(pk)
     }
+
+    /// Whether or not this key has multiple derivation paths.
+    pub fn is_multipath(&self) -> bool {
+        match *self {
+            DescriptorSecretKey::Single(..) | DescriptorSecretKey::XPrv(..) => false,
+            DescriptorSecretKey::MultiXPrv(_) => true,
+        }
+    }
```
Comment (darosior) [inline src/descriptor/key.rs:324]: This seems redundant with `MiniscriptKey`'s `num_der_paths` introduced later on, but i kept it here in order to also have it for private keys.

**Comment by darosior** [comment]
Rebased on master following the recent merges for the 8.0.0 release.

**Review by darosior — COMMENTED** [review]
I've integrated this in my downstream project. I found a bug for which i'm pushing the fix and a unit test triggering it. Aside from this bug i've successfully used it and by doing so cross-checked it against the Python implementation i'm using in the functional testing framework of my software (https://github.com/darosior/python-bip380/pull/21).

**Comment by darosior** [comment]
Got a spurious io issue on the CPP tests:
```
thread 'tests_from_cpp' panicked at 'called `Result::unwrap()` on an `Err` value: JsonRpc(Transport(SocketError(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" })))', tests/test_cpp.rs:80:10
```

Force-pushing to trigger a re-run.

**Comment by darosior** [comment]
I hadn't noticed the `key` submodule was made private. I just pushed a patch to re-export `DescriptorMultiXKey` like `DescriptorXKey`. H/T @edouardparis for mentioning that to me.

**Comment by darosior** [comment]
Friendly ping to know when you'll be able to review this, or if you think it is blocked by something?

**Comment by sanket1729** [comment]
@darosior, this is on my list for this week.

File: src/descriptor/mod.rs [hunk 2]
```diff
@@ -792,6 +792,42 @@ impl Descriptor<DescriptorPublicKey> {
     }
 }
 
+impl<Pk: MiniscriptKey> Descriptor<Pk> {
+    /// Whether this descriptor is a multipath descriptor that contains any 2 multipath keys
+    /// with a different number of derivation paths.
+    /// Such a descriptor is invalid according to BIP389.
+    pub fn multipath_length_mismatch(&self) -> bool {
```
Comment (sanket1729) [inline src/descriptor/mod.rs:799]: Just a review comment. Nothing to be addressed here. Ideally, we want to make sure that such invalid descriptors cannot be created in the first place. Adding this pub might confuse users to call it to double-check everything. 

I think this is true for all APIs that deal with Miniscript, but not for the internal `Terminal`. I have some changes in mind in future PRs to remove all the confusing sanity_check/check APIs.

**Review by sanket1729 — COMMENTED** [review]
done reviewing a63d5a263a9006b4d29342012133a3bc919765ba. Just Left a couple of comments. The rest of it is good to go. I have some other changes in mind to come in the future that would clean up the translatePk trait so which would make it impossible to create Miniscripts with different multi-path lengths even after translating. 

Right now, it would fall upon the user to explicitly call the mismatching API to make sure that the translation is done correctly.

File: src/descriptor/key.rs [hunk 3]
```diff
@@ -61,6 +65,19 @@ pub struct DescriptorXKey<K: InnerXKey> {
...
 
+/// Instance of one or more extended keys, as specified in BIP 389.
+#[derive(Debug, Eq, PartialEq, Clone, Ord, PartialOrd, Hash)]
+pub struct DescriptorMultiXKey<K: InnerXKey> {
+    /// Origin information
+    pub origin: Option<(bip32::Fingerprint, bip32::DerivationPath)>,
+    /// The extended key
+    pub xkey: K,
+    /// The derivation paths. Never empty.
+    pub derivation_paths: Vec<bip32::DerivationPath>,
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/descriptor/key.rs:76]: Heh i was on the fence about this when i wrote it. Guess that settles it. :) I added a `DerivPaths` newtype.

**Review by sanket1729 — APPROVED** [review]
ACK ae0f3d95fa408bfc5e535050743a0f0ed8cb2963.

**Comment by sanket1729** [comment]
@apoelstra, any thoughts on merging this? This is not a merged BIP yet and could be changed. But so are many things in this repo that are not even BIP yet.

**Pull request merged**